### DISCUSSION TITLE
Liquidsoap in Docker Containers

### DISCUSSION THREAD

**Discussion by gAlleb** [savonet/liquidsoap#3932] [General]
https://github.com/savonet/liquidsoap/discussions/3925#discussioncomment-9508766

Thank you, @vitoyucepi!

> I don't know why you would use alpine:edge as a base image for a container, but I would prefer ubuntu instead.

I've just looked here: https://github.com/savonet/liquidsoap/blob/main/.github/docker/Dockerfile.production-alpine
And if I try to do the same with older versions I get 
```ruby
1.959 fetch https://dl-cdn.alpinelinux.org/alpine/v3.19/community/x86_64/APKINDEX.tar.gz
3.811 ERROR: unable to select packages:
3.822   so:libswresample.so.4 (no such package):
3.822     required by: liquidsoap-x86_64-2.2.5-r0[so:libswresample.so.4]
```
Main idea was keeping image minimal but with python3 and ffprobe.

> Here are parts of my repo. All files are licensed under the GNU AGPLv3.

Looks nice, but some parts are not understandable. Why do you use both `alpine` and `ubuntu`? Alpine just to download the package?
I think I can build image with `ubuntu` but this would require super skills to know how to install `only needed` packages for LS .deb and `cue_file` to work. Because if one installs blindly - it would all take up to 1GB of space again.

> `entrypoint`

This bit seems distant as ever.

**Comment by vitoyucepi** [comment]
> Looks nice, but some parts are not understandable. Why do you use both alpine and ubuntu? Alpine just to download the package?

Yes, my `containerfile` uses alpine as the downloader. I used to use `COPY` to transfer files between stages, but that results in an additional non-removable layer.

> it would all take up to 1GB of space again.

It won't, only 0.5 GB. Try to build it.
If you want a smaller image, you'll have to build liquidsoap and ffmpeg from source to disable all unused features.

**Reply by gAlleb** [reply]
Thanks, Vito.

I didn't understand 

```
COPY rootfs /

ENTRYPOINT [ "entrypoint" ]
CMD [ "liquidsoap", "/var/liquidsoap/default.liq" ]
```

and bash script completly. If you you could enlighten me what does bash script do and why do we need `CMD [ "liquidsoap", "/var/liquidsoap/default.liq" ]` ?. `COPY rootfs /` always  showed an error.

So I switched to `ENTRYPOINT ["/usr/bin/tini", "--", "/usr/bin/liquidsoap"]`

And the container is 602MB. I added ffmpeg and python an tini. And it's now 632MB. And It works.

```ruby
FROM alpine:3.19 as downloader

ARG LIQUIDSOAP_VERSION=2.2.5

ARG LIQUIDSOAP_RELEASE="https://github.com/savonet/liquidsoap-release-assets/releases/download/v${LIQUIDSOAP_VERSION}"
ARG LIQUIDSOAP_DEB="liquidsoap_${LIQUIDSOAP_VERSION}-ubuntu-jammy-1_amd64.deb"
ARG LIQUIDSOAP_SHA512SUM="56a39e326019ca6f8f25017407f4edac2b1d0b18995e9fba4558e566050c9d48513ab1dbff56034a4ddcc2bed814d51437e400942639605784d2c0e189cc2e0e"

RUN set -eux; \
      mkdir /packages; \
      wget "${LIQUIDSOAP_RELEASE}/${LIQUIDSOAP_DEB}" -O /packages/liquidsoap.deb; \
      echo "${LIQUIDSOAP_SHA512SUM}  /packages/liquidsoap.deb" | sha512sum -c -;

FROM ubuntu:22.04

ENV TZ=UTC

ARG CA_CERTIFICATES_VERSION=20230311ubuntu0.22.04.1
ARG GOSU_VERSION=1.14-1
ARG TZDATA_VERSION=2024a-0ubuntu0.22.04

RUN --mount=type=bind,target=/packages,source=/packages,from=downloader \
    set -eux; \
      apt-get update; \
      apt-get install -y --no-install-recommends \
        ca-certificates="${CA_CERTIFICATES_VERSION}" \
	tini \
	python3 \
        gosu="${GOSU_VERSION}" \
        tzdata="${TZDATA_VERSION}" \
        /packages/liquidsoap.deb \
	ffmpeg \
      ;

ENTRYPOINT ["/usr/bin/tini", "--", "/usr/bin/liquidsoap"]
```

And also I would like to ask why `alpine image` which I made doesn't work?

**Reply by vitoyucepi** [reply]
1. Create the `rootfs` subdirectory in the build root.
   Place the `entrypoint` script in `rootfs/usr/bin`.
1. Do you need `tini`? I like to use the `entrypoint` script instead.
1. Why can't you just `FROM savonet/liquidsoap-alpine:v2.2.5`?

**Reply by gAlleb** [reply]
> Create the rootfs subdirectory in the build root.
Place the entrypoint script in rootfs/usr/bin.
Do you need tini? I like to use the entrypoint script instead.

Oh, I see. I just couldn't make out how to treat that script.

> Why can't you just FROM savonet/liquidsoap-alpine:v2.2.5?

I can. And it works. And it's build with edge.

Here https://hub.docker.com/layers/savonet/liquidsoap-alpine/v2.2.5/images/sha256-0ef1016f66341cba94220b1a3f25c6f9182f66610d1248a2094e624aff6e6f0d?context=explore

``` ruby
RUN |1 APK_FILE=artifacts/c3f311e/alpine-packages/liquidsoap-x86_64-2.2.5-r0.apk /bin/sh -c apk add --allow-untrusted --no-cache       -X http://dl-cdn.alpinelinux.org/alpine/edge/testing       tini /tmp/liquidsoap.apk &&       rm -rf /tmp/liquidsoap.apk # buildkit
```

It's build somehow and it's a working image. But I need python3 and ffmpeg (ffprobe) to accomodate `cue_file` for `autocue`.

I even tried the following: installed into it python3 and ffmpeg and saved as new image and it stopped working: LS started showing errors.

**Reply by vitoyucepi** [reply]
```containerfile
FROM savonet/liquidsoap-alpine:v2.2.5

USER root

RUN apk add --no-cache ffmpeg python3

USER liquidsoap
```

> LS started showing errors.

Did I miss something?

**Reply by gAlleb** [reply]
Ha) You see, that's the difference between someone who started learning docker few days ago and you)

It's elegant and it works! And I have tried doing it by modifying container by hand (with `--entrypoint sh`). :)

Thanks so much, Vito!

> Why can't you just FROM savonet/liquidsoap-alpine:v2.2.5?

I got it now. First time I didn't) Thanks.

But why github actions can produce it with edge?

**Reply by vitoyucepi** [reply]
> But why github actions can produce it with edge?

Why can github actions create container image with *alpine:edge* as base image, but you can't?

```containerfile
FROM alpine:edge as downloader

ARG LIQUIDSOAP_VERSION=2.2.5

ARG LIQUIDSOAP_RELEASE="https://github.com/savonet/liquidsoap-release-assets/releases/download/v${LIQUIDSOAP_VERSION}"
ARG LIQUIDSOAP_APK="liquidsoap-x86_64-${LIQUIDSOAP_VERSION}-r0.apk"
ARG LIQUIDSOAP_SHA512SUM="0d433e902b296cc53f26d52c3805c1880b3a916898b4ac709120017803e787c57530974fab4e4b33df0237f70f0f40ecb1ad91e9bcedfad0cccafe8a37edcedc"

RUN set -eux; \
      mkdir /packages; \
      wget "${LIQUIDSOAP_RELEASE}/${LIQUIDSOAP_APK}" -O /packages/liquidsoap.apk; \
      echo "${LIQUIDSOAP_SHA512SUM}  /packages/liquidsoap.apk" | sha512sum -c -;

FROM alpine:edge

RUN --mount=type=bind,target=/packages,source=/packages,from=downloader \
    apk add --no-cache --allow-untrusted \
      -X https://dl-cdn.alpinelinux.org/alpine/edge/testing \
      /packages/liquidsoap.apk \
      ffmpeg \
      python3 \
    ;

USER liquidsoap
```

**Reply by gAlleb** [reply]
Thank you, Vito! I built it! 

Just for testing purposes - without entrypoint:

```
Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "/home/radio/liquidsoap/omfm/index.liq": permission denied: unknown
```

And with `ENTRYPOINT ["/sbin/tini" "--" "/usr/bin/liquidsoap"]` and apk tini
And it's the same thing as I wrote in this comment https://github.com/savonet/liquidsoap/discussions/3925#discussioncomment-9508262 
LS behaves like this:

```
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
```

**Reply by vitoyucepi** [reply]
Run into the interactive shell and test the script by hand.
`liquidsoap /home/radio/liquidsoap/omfm/index.liq`

**Reply by gAlleb** [reply]
```
 liquidsoap /home/radio/liquidsoap/omfm/index.liq
At line 1, char 0:

Error 3: Syntax error!
```

with

```
FROM savonet/liquidsoap-alpine:v2.2.5
USER root
RUN apk add --no-cache ffmpeg python3
USER liquidsoap
```

All is superb.

**Reply by vitoyucepi** [reply]
Check the permissions for `/home/radio/liquidsoap/omfm/index.liq`.
Or just `COPY` the scripts at the build stage.

**Reply by gAlleb** [reply]
I've got everywhere 644 and 755 (radio:radio, 1001:1001)

Only this image refuses to work. Every other one work fine.

**Reply by gAlleb** [reply]
Created with COPY

```
radio@sxvx:~$ docker run -it --entrypoint sh galleb/liquidsoap-alpine-pf:0.2
/ $ ls
bin    dev    etc    home   lib    media  mnt    opt    proc   root   run    sbin   srv    sys    tmp    usr    var
/ $ cd home
/home $ ls
radio
/home $ cd radio
/home/radio $ cd omfm
sh: cd: can't cd to omfm: No such file or directory
/home/radio $ cd liquidsoap/omfm
/home/radio/liquidsoap/omfm $ ls
index.liq  omfm.liq
/home/radio/liquidsoap/omfm $ liquidsoap index.liq
At index.liq, line 13, char 88-88:
%include "/home/radio/liquidsoap/autocue-integrate-with-liquidsoap/autocue.cue_file.liq"

Error 14: Uncaught runtime error:
type: not_found,
message: "File not found!",
stack: at index.liq, line 13, char 88-88

/home/radio/liquidsoap/omfm $ 
```

So it works inside the container.  Only if I COPY - but it's nonsense. Something isn't right here. Next step - COPY entire `radio` directory)))


UPDATE! Yeah. With this build LS works only inside and nothing could be mounted.

Every other build works fine. 

```
FROM savonet/liquidsoap-alpine:v2.2.5
USER root
RUN apk add --no-cache ffmpeg python3
USER liquidsoap
```
This works fine!

Everything. Except handmade alpine:edge

**Reply by vitoyucepi** [reply]
I'd like to see if running container like this
`docker run -it --rm --network host -v '/home/radio/liquidsoap/omfm:/tmp/test' -w '/tmp/test' --entrypoint sh image_name:image_tag`
and then
`liquidsoap index.liq`
actually works.
At least mount `'/home/radio/liquidsoap:/tmp/test'`.

If it doesn't show me the `stat index.liq` and `stat /tmp/test`.

**Reply by gAlleb** [reply]
```ruby
radio@sxvx:~$ docker run -it --rm --network host -v '/home/radio/liquidsoap/omfm:/tmp/test' -w '/tmp/test' --entrypoint sh galleb/liquidsoap-alpine-wtf:v01
Unable to find image 'galleb/liquidsoap-alpine-wtf:v01' locally
v01: Pulling from galleb/liquidsoap-alpine-wtf
4edfc05e3af2: Already exists 
ee11cbb98958: Pull complete 
Digest: sha256:581968be94fbb666f9ff785ecd2572baf415c2f6fa9fa354a23c4c7fc37900b9
Status: Downloaded newer image for galleb/liquidsoap-alpine-wtf:v01
/tmp/test $ liquidsoap index.liq
At index.liq, line 13, char 88-88:
%include "/home/radio/liquidsoap/autocue-integrate-with-liquidsoap/autocue.cue_file.liq"

Error 14: Uncaught runtime error:
type: not_found,
message: "File not found!",
stack: at index.liq, line 13, char 88-88

/tmp/test $ stat index.liq
  File: index.liq
  Size: 2004      	Blocks: 8          IO Block: 4096   regular file
Device: fd01h/64769d	Inode: 320827      Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1001/ UNKNOWN)   Gid: ( 1001/ UNKNOWN)
Access: 2024-05-21 21:12:11.205175761 +0000
Modify: 2024-05-20 20:49:07.817537392 +0000
Change: 2024-05-20 20:49:07.958537702 +0000
/tmp/test $ stat /tmp/test
  File: /tmp/test
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: fd01h/64769d	Inode: 320794      Links: 3
Access: (0775/drwxrwxr-x)  Uid: ( 1001/ UNKNOWN)   Gid: ( 1001/ UNKNOWN)
Access: 2024-05-21 21:14:41.178944010 +0000
Modify: 2024-05-19 20:09:43.673152138 +0000
Change: 2024-05-19 20:09:43.673152138 +0000
/tmp/test $ ls
index.liq            log                  omfm.liq             request.dynamic.liq
/tmp/test $ 
```


it works. But what's the point?
It sees even index.liq (I didn't COPY it) and all other files. So it mounts.

**Reply by vitoyucepi** [reply]
I think this behavior is intentional, because all components of the path should be accessible by the user liquidsoap.
Imagine you have `/home/radio` with `0700`, but `/home/radio/liquidsoap/index.liq` has `0644`. I don't think a non-owner should be able to read `index.liq` in this case.

For example, I `COPY` all the liquidsoap scripts during the image build phase, but I mount volatile files like music during the deploy phase.

**Reply by gAlleb** [reply]
Well, hard to say. If it's intentional. Why other images work? And I specifically pass the user (1001:1001) so it could read and even wright.

**Reply by vitoyucepi** [reply]
Mount the volume as before and try to view it with the following script
`find /home/radio -name '*.liq' -exec stat -c "%A %U:%G %n" {} \;`
This should find all the files you need. If you have any problems, try running the container with the `-u root` argument.

**Reply by gAlleb** [reply]
I've tried `docker run -it --rm --network host -v '/home/radio:/home/radio' -w '/home/radio' --entrypoint sh galleb/liquidsoap-alpine-wtf:v01` and it sees all files. And starts the script till permission for logs are denied.

But with
```
services:
  liquidsoap:
    image: galleb/liquidsoap-alpine-wtf:v01
    command:
      - /home/radio/liquidsoap/omfm/index.liq
    volumes:
      - /home/radio:/home/radio
    network_mode: "host"
    restart: always
```

I get
```
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
liquidsoap-1  | /home/radio/liquidsoap/omfm/index.liq: line 0: [/sbin/tini: not found
```

**Reply by vitoyucepi** [reply]
And again, why do you need `tini` at all? Use `liquidsoap` as the entrypoint.

**Reply by gAlleb** [reply]
I don't) I learn.

You mean use you script while building?

You think then it would show more signs of life?

**Reply by gAlleb** [reply]
Ha) Vito, It's Marvelous!

So I

```ruby
FROM alpine:edge as downloader

ARG LIQUIDSOAP_VERSION=2.2.5

ARG LIQUIDSOAP_RELEASE="https://github.com/savonet/liquidsoap-release-assets/releases/download/v${LIQUIDSOAP_VERSION}"
ARG LIQUIDSOAP_APK="liquidsoap-x86_64-${LIQUIDSOAP_VERSION}-r0.apk"
ARG LIQUIDSOAP_SHA512SUM="0d433e902b296cc53f26d52c3805c1880b3a916898b4ac709120017803e787c57530974fab4e4b33df0237f70f0f40ecb1ad91e9bcedfad0cccafe8a37edcedc"

RUN set -eux; \
      mkdir /packages; \
      wget "${LIQUIDSOAP_RELEASE}/${LIQUIDSOAP_APK}" -O /packages/liquidsoap.apk; \
      echo "${LIQUIDSOAP_SHA512SUM}  /packages/liquidsoap.apk" | sha512sum -c -;

FROM alpine:edge
ENV TZ=UTC
RUN --mount=type=bind,target=/packages,source=/packages,from=downloader \
    apk add --no-cache --allow-untrusted \
      -X https://dl-cdn.alpinelinux.org/alpine/edge/testing \
      /packages/liquidsoap.apk \
      ffmpeg \
      python3 \
      tzdata \
    ;

USER liquidsoap

COPY rootfs /

ENTRYPOINT [ "entrypoint" ]
```
And it works with my docker compose! Ha!

So `entrypoint` turns out to be an important thing!

Thanks! 
Tell me more! I'm kidding) But do.

**Reply by vitoyucepi** [reply]
Just one more thing.
1. If you use my `entrypoint`, you don't need to define `USER liquidsoap`, but you do need to add `su-exec`. Here is the `entrypoint` script for alpine. The license is GNU AGPLv3.
   ```sh
   #!/usr/bin/env sh

   set -eu

   main() {
     if [ "${1#-}" != "$1" ] || [ "${1%.liq}" != "$1" ]; then
       set -- liquidsoap "$@"
     fi

     if [ "$1" = 'liquidsoap' ] && [ "$(id -u)" -eq '0' ]; then
       exec su-exec liquidsoap "$0" "$@"
     fi

     exec "$@"
   }

   main "$@"
   ```

1. To test things, you don't even need any of these commands, just set `ENTRYPOINT [ "liquidsoap" ]` and use `CMD` or command to select the script.

1. If you do http requests, then you should also install `ca-certificates`.

**Reply by gAlleb** [reply]
I see.

> If you do http requests, then you should also install ca-certificates.

I do. They work now, but they are local. I presume it's for external ones? For alpine it enough just to install `ca-certificates` without any version?

Tnanks!

**Reply by vitoyucepi** [reply]
`ca-certificates` is the list of trusted root and intermediate certificates. You need them to validate secure connections.
You may want to pin the version, but it's not necessary.

**Reply by gAlleb** [reply]
I see. Thanks so much! Gonna dig further!

**Comment by Moonbase59** [comment]
@gAlleb Remember you’ll need an ffmpeg with `libswresample` for `cue_file` to work.

**Reply by vitoyucepi** [reply]
To use liquidsoap with alpine:3.19, you need to rebuild the binary.

- https://pkgs.alpinelinux.org/contents?file=&path=&name=ffmpeg-libswresample&branch=v3.19&arch=x86_64
- https://pkgs.alpinelinux.org/contents?file=&path=&name=ffmpeg-libswresample&branch=edge&&arch=x86_64

**Reply by gAlleb** [reply]
So it won't work with `edge`?

**Reply by vitoyucepi** [reply]
Without rebuilding it will not work with *3.19*.
The main problem with *alpine:edge* is the rolling release. Ref: #3875.

**Comment by gAlleb** [comment]
Hi, @vitoyucepi. So I found another use for Docker :) and compiled `centrifugo` binary inside the container

```ruby
FROM golang:1.22-bookworm 

RUN apt-get update \
    && apt-get install -y --no-install-recommends openssl git

RUN go install github.com/jwilder/dockerize@v0.6.1

RUN go install github.com/aptible/supercronic@v0.2.28

RUN mkdir -p /tmp/centrifugo \
    && cd /tmp/centrifugo \
    && git clone https://github.com/centrifugal/centrifugo.git . \
    && git checkout 5.4.0_dev \
    && go build \
    && mv ./centrifugo /go/bin/centrifugo
```

It's nice so you don't have to install all the packages on your system and increase number of trash. Then I got a 60MB binary and removed 2GB! container.

Binary is working fine. So I decided to make a docker image with it to test (and maybe accommodating it within LS Docker). It is straightforward and I've made the same steps as guys at centrifugo do (alike their docker repository):

```ruby
FROM alpine:3.19

RUN apk --no-cache upgrade && apk --no-cache add ca-certificates  

COPY centrifugo /usr/local/bin/centrifugo

WORKDIR /centrifugo

CMD ["centrifugo"]
```

But It refuses to work (and I have tried even pulling their official binary and building again with it). Just doesn't want to start:

```
exec /usr/local/bin/centrifugo: no such file or directory
```

That's what I got. I've tried lot's of options, with x, a+x, non of that too — doesn't work. Any insights on that? Probably I'm doing something wrong. Thanks!

**Reply by vitoyucepi** [reply]
Why not just use the prebuild image https://hub.docker.com/r/centrifugo/centrifugo?

> `RUN apk --no-cache upgrade`

I can't recommend this because it may have unknown side effects.

> `FROM golang:1.22-bookworm`
> `FROM alpine:3.19`

I'm not 100% sure about go, but here are my thoughts.
The go binaries built in Debian expect `glibc`. Alpine can only provide `musl`.
Either use Alpine in both cases, or `CGO_ENABLED=0`.

**Reply by gAlleb** [reply]
> Why not just use the prebuild image https://hub.docker.com/r/centrifugo/centrifugo?
    RUN apk --no-cache upgrade

It's an unreleased version with some new features like `cached recovery`. So no need for `on connect backend proxy server` (https://github.com/savonet/liquidsoap/discussions/3782#discussioncomment-8771678). Last event is sent on connect from history with just couple of lines added to centrifugo config.



>  I'm not 100% sure about go, but here are my thoughts.
The go binaries built in Debian expect glibc. Alpine can only provide musl.
Either use Alpine in both cases, or CGO_ENABLED=0.

I don't thinks that this is related. Because I've tried the following:
1. Pull official image
2. Extract centrifugo binary
3. Make new image with that binary with the same OS version and steps as per https://github.com/centrifugal/centrifugo/blob/master/Dockerfile
4. Got the same results: `exec /usr/local/bin/centrifugo: no such file or directory`

 

Probably my mistake somewhere and I am missing something.

**Reply by vitoyucepi** [reply]
```containerfile
# FROM golang:1.22.3-alpine3.20 as build
FROM golang:1.22.3-bookworm as build

# RUN apk add git
RUN set -eux; \
      apt-get update; \
      apt-get install --no-install-recommends git;

RUN git clone https://github.com/centrifugal/centrifugo.git -b 5.4.0_dev --depth 1

WORKDIR /go/centrifugo

# RUN go build
RUN CGO_ENABLED=0 go build

FROM alpine:3.20

COPY --from=build /go/centrifugo/centrifugo /usr/bin
RUN centrifugo version
```

**Reply by gAlleb** [reply]
That is great. Well, live and learn. Thank you very much, Vito.
